spring-integration-zookeeper/src/test/java/org/springframework/integration/zookeeper/metadata/ZookeeperMetadataStoreTests.java

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.zookeeper.metadata;
+
+import static org.hamcrest.collection.IsCollectionWithSize.hasSize;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+import static org.springframework.integration.zookeeper.metadata.EqualsResultMatcher.equalsResult;
+import static org.springframework.integration.zookeeper.metadata.EventuallyMatcher.eventually;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import com.gs.collections.impl.factory.Lists;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.CuratorFrameworkFactory;
+import org.apache.curator.retry.BoundedExponentialBackoffRetry;
+import org.apache.curator.test.TestingServer;
+import org.hamcrest.collection.IsIterableContainingInOrder;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import org.springframework.beans.DirectFieldAccessor;
+import org.springframework.integration.zookeeper.metadata.EqualsResultMatcher.Evaluator;
+
+/**
+ * @author Marius Bogoevici
+ */
+public class ZookeeperMetadataStoreTests {
+
+	private static final Log log = LogFactory.getLog(ZookeeperMetadataStore.class);
+
+	private TestingServer testingServer;
+
+	private CuratorFramework client;
+
+	private ZookeeperMetadataStore metadataStore;
+
+	@Before
+	public void setUp() throws Exception{
+		testingServer = new TestingServer(true);
+		client = createNewClient();
+		metadataStore = new ZookeeperMetadataStore(client);
+		metadataStore.start();
+	}
+
+	@After
+	public void tearDown() {
+		this.metadataStore.stop();
+		closeClient(this.client);
+		try {
+			testingServer.stop();
+		}
+		catch (IOException e) {
+			log.warn("Exception thrown while shutting down ZooKeeper: ", e);
+		}
+		testingServer.getTempDirectory().delete();
+	}
```

Comment:
Can't we move all this stuff to the `static` class-level initialization phase to increase tests performance? 
I've done that locally and see a race condition with listeners test. But it is expected, because with that `metadataStore` is shared.
Of course, we can live with those `20 sec` and can revise in the future.
Just in case to share with you a sign...